pandas/core/arrays/period.py

```diff
@@ -733,9 +733,13 @@ def asfreq(self, freq=None, how: str = "E") -> Self:
         '2015-01'], dtype='period[M]')
         """
         how = libperiod.validate_end_alias(how)
+
+        freq_original = freq
         if isinstance(freq, BaseOffset):
             freq = freq_to_period_freqstr(freq.n, freq.name)
         freq = Period._maybe_convert_freq(freq)
+        if not hasattr(freq, "_period_dtype_code"):
+            raise TypeError(f'"{freq_original}" is not supported as a period frequency')
```

Comment:
in https://github.com/pandas-dev/pandas/pull/56945#pullrequestreview-1854233737 I'm suggesting to raise if the return value of `to_offset(..., is_period=True)` doesn't have a `_period_dtype_code` attribute

If that works, then it would resolve the linked issue too because `Period._maybe_convert_freq` would go down that path